### ISSUE TITLE
OOM killer and memory leak while running sysbench

### ISSUE THREAD

**Issue by mdcallag** [orioledb/orioledb#276]
I am unable to avoid the OOM killer while running sysbench. I assume the issue is memory leaks rather than just using a lot more memory than expected.

My server is [described here](https://smalldatum.blogspot.com/2022/10/small-servers-for-performance-testing-v4.html) and has 8 cores with 16G of RAM. The server uses Ubuntu 22.04.

I compiled the OracleDB fork of Postgres at patches16_18 and then the OrioleDB extension at https://github.com/orioledb/orioledb/commit/bbfc8ab2ba6381334a871c7fcbf787feb723041f as [described here](https://github.com/orioledb/orioledb/issues/275).

Then I run sysbench via my (messy) helper scripts. The scripts run ~43 microbenchmarks and each microbenchmark starts a new sysbench process and gets new connections to Postgres. Postgres is initialized at the start and only shutdown after the 43 microbenchmarks finish. 

There were several rounds of this and all appear to trigger the OOM killer to kill Postgres processes. Round 1 used the default config for OrioleDB, I didn't set any options. I also used prepared statements with sysbench. Read-only microbenchmarks ran for 10 minutes, non read-only for 20 minutes. The OOM occurs on the ~10th microbenchmark.

Round 2 set options for OrioleDB including main_buffers=8GB. I used sysbench with prepared statements. There was OOM again on a microbenchmark that did range queries. The options were:
```
orioledb.main_buffers = 8GB
orioledb.catalog_buffers = 64MB
orioledb.undo_buffers = 256MB
orioledb.recovery_queue_size = 64MB
orioledb.checkpoint_completion_ratio = 1.0
orioledb.bgwriter_num_workers = 3
```

Round 3 reduced main_buffers to 1GB and otherwise used the options above. Sysbench used prepared statements. There was OOM again on a microbenchmark that did range queries.

Round 4 also used main_buffers=1GB and the options above. Sysbench did not use prepared statements. There was OOM again, much faster than the cases above, on a microbenchmark that did point queries.

**Comment by mdcallag** [comment]
From round 4, here is output from ps running in a loop for the Postgres processes for my connections. Column 6 is RSS. After CREATE INDEX it quickly grows to 15134564 (~15G).

```
mdcallag 3035163 84.5  2.2 9366648 348380 ?      Ss   23:27   0:06 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 84.1  3.2 9367928 505820 ?      Rs   23:27   0:15 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 84.1  4.2 9375096 668920 ?      Rs   23:27   0:23 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.8  5.2 9375096 820804 ?      Rs   23:27   0:31 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.8  6.2 9375096 978664 ?      Rs   23:27   0:40 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.9  7.1 9375096 1133252 ?     Rs   23:27   0:48 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.7  8.1 9375096 1287840 ?     Rs   23:27   0:56 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.8  9.1 9375096 1439444 ?     Rs   23:27   1:05 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.8 10.1 9375096 1593044 ?     Rs   23:27   1:13 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 82.9 11.0 9367160 1741284 ?     Rs   23:27   1:22 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.0 12.0 9367160 1895484 ?     Rs   23:27   1:30 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.1 12.9 9366648 2050032 ?     Ss   23:27   1:38 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.1 14.0 9375096 2209724 ?     Rs   23:27   1:47 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.2 14.9 9375096 2362444 ?     Rs   23:27   1:55 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.2 15.9 9366648 2509448 ?     Ss   23:27   2:04 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.3 16.8 9367928 2663220 ?     Rs   23:27   2:12 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.3 17.8 9375096 2818520 ?     Rs   23:27   2:20 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.4 18.8 9366648 2965936 ?     Ss   23:27   2:29 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.4 19.8 9375096 3125552 ?     Rs   23:27   2:37 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.4 20.7 9366648 3270608 ?     Ss   23:27   2:46 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.4 21.7 9371000 3426636 ?     Rs   23:27   2:54 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 22.7 9375096 3585828 ?     Rs   23:27   3:02 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 23.6 9375096 3737408 ?     Rs   23:27   3:11 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 24.6 9375096 3890384 ?     Rs   23:27   3:19 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 25.6 9375096 4043472 ?     Rs   23:27   3:28 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 26.5 9375096 4192036 ?     Rs   23:27   3:36 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.5 27.5 9375096 4341988 ?     Rs   23:27   3:44 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.6 28.5 9375096 4495448 ?     Rs   23:27   3:53 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 83.6 29.4 9366648 4639552 ?     Ss   23:27   4:01 postgres: root ib 127.0.0.1(39020) idle
mdcallag 3035163 83.6 30.3 9371000 4793932 ?     Rs   23:27   4:10 postgres: root ib 127.0.0.1(39020) INSERT
mdcallag 3035163 84.0 31.6 9530432 4990868 ?     Rs   23:27   4:19 postgres: root ib 127.0.0.1(39020) CREATE INDEX
mdcallag 3035163 84.5 32.7 9697724 5158772 ?     Rs   23:27   4:29 postgres: root ib 127.0.0.1(39020) CREATE INDEX
mdcallag 3035163 84.9 32.7 9701852 5162732 ?     Rs   23:27   4:39 postgres: root ib 127.0.0.1(39020) CREATE INDEX
mdcallag 3035495 88.8 19.3 10347596 3052336 ?    Rs   23:32   0:04 postgres: root ib 127.0.0.1(47374) SELECT
mdcallag 3035495 87.3 42.7 12304628 6736856 ?    Rs   23:32   0:13 postgres: root ib 127.0.0.1(47374) SELECT
mdcallag 3035495 87.4 59.0 14319852 9310864 ?    Rs   23:32   0:21 postgres: root ib 127.0.0.1(47374) SELECT
mdcallag 3035495 87.2 72.8 16325672 11484748 ?   Rs   23:32   0:30 postgres: root ib 127.0.0.1(47374) idle
mdcallag 3035495 87.0 85.0 18203856 13409492 ?   Rs   23:32   0:39 postgres: root ib 127.0.0.1(47374) SELECT
mdcallag 3035495 85.9 96.0 20041192 15142572 ?   Rs   23:32   0:47 postgres: root ib 127.0.0.1(47374) idle
mdcallag 3035495 83.3 95.9 21431884 15134564 ?   Rs   23:32   0:54 postgres: root ib 127.0.0.1(47374) SELECT
mdcallag 3035495 76.4 96.0 21839140 15146904 ?   Ds   23:32   0:57 postgres: root ib 127.0.0.1(47374) idle 
```

**Comment by mdcallag** [comment]
From round 3, it doesn't immediately get OOM after the CREATE INDEX, but further along I see ...

```
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   7:41 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   7:51 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:00 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:10 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:19 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:29 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Ss   23:05   8:38 postgres: root ib 127.0.0.1(38888) idle
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:48 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   8:57 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   9:07 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   9:16 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3033906 94.6 29.1 9366112 4599364 ?     Rs   23:05   9:26 postgres: root ib 127.0.0.1(38888) SELECT
mdcallag 3034497 85.5 10.4 9536216 1642392 ?     Rs   23:15   0:01 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 79.5 31.8 10342580 5016784 ?    Rs   23:15   0:09 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 78.7 39.9 11160904 6301968 ?    Rs   23:15   0:17 postgres: root ib 127.0.0.1(37150) BIND
mdcallag 3034497 78.7 45.7 11992984 7212536 ?    Rs   23:15   0:25 postgres: root ib 127.0.0.1(37150) idle
mdcallag 3034497 78.7 51.1 12829104 8063696 ?    Ss   23:15   0:33 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 78.5 56.3 13651736 8887744 ?    Rs   23:15   0:40 postgres: root ib 127.0.0.1(37150) BIND
mdcallag 3034497 78.5 61.5 14478988 9715080 ?    Rs   23:15   0:48 postgres: root ib 127.0.0.1(37150) BIND
mdcallag 3034497 78.5 66.9 15318208 10554312 ?   Rs   23:15   0:56 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 78.4 72.1 16139520 11375784 ?   Rs   23:15   1:04 postgres: root ib 127.0.0.1(37150) idle in transaction
mdcallag 3034497 78.4 77.3 16969232 12205488 ?   Rs   23:15   1:12 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 78.5 82.6 17807776 13042416 ?   Rs   23:15   1:20 postgres: root ib 127.0.0.1(37150) idle in transaction
mdcallag 3034497 78.4 87.9 18631476 13864828 ?   Rs   23:15   1:27 postgres: root ib 127.0.0.1(37150) idle in transaction
mdcallag 3034497 78.3 93.2 19467204 14700432 ?   Rs   23:15   1:35 postgres: root ib 127.0.0.1(37150) idle in transaction
mdcallag 3034497 78.4 95.2 20285276 15024212 ?   Rs   23:15   1:43 postgres: root ib 127.0.0.1(37150) BIND
mdcallag 3034497 78.4 95.1 21082872 15005564 ?   Ss   23:15   1:51 postgres: root ib 127.0.0.1(37150) COMMIT
mdcallag 3034497 78.0 95.2 21804008 15021284 ?   Rs   23:15   1:58 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 73.5 95.2 21804360 15022148 ?   Ds   23:15   1:59 postgres: root ib 127.0.0.1(37150) SELECT
mdcallag 3034497 69.9 95.2 21804556 15022656 ?   Ds   23:15   2:01 postgres: root ib 127.0.0.1(37150) SELECT
```

**Comment by mdcallag** [comment]
OOM killer messages from dmesg

```
[Mon Sep 25 20:30:05 2023] postgres invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Mon Sep 25 20:30:05 2023]  oom_kill_process.cold+0xb/0x10
[Mon Sep 25 20:30:05 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Mon Sep 25 20:30:05 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=2971789,uid=1000
[Mon Sep 25 20:30:05 2023] Out of memory: Killed process 2971789 (postgres) total-vm:30671928kB, anon-rss:14987048kB, file-rss:1280kB, shmem-rss:79560kB, UID:1000 pgtables:39156kB oom_score_adj:0
[Mon Sep 25 20:57:39 2023] systemd invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Mon Sep 25 20:57:39 2023]  oom_kill_process.cold+0xb/0x10
[Mon Sep 25 20:57:39 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Mon Sep 25 20:57:39 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=init.scope,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=2973003,uid=1000
[Mon Sep 25 20:57:39 2023] Out of memory: Killed process 2973003 (postgres) total-vm:29937612kB, anon-rss:15027568kB, file-rss:1048kB, shmem-rss:99000kB, UID:1000 pgtables:37728kB oom_score_adj:0
[Tue Sep 26 14:00:14 2023] postgres invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Tue Sep 26 14:00:14 2023]  oom_kill_process.cold+0xb/0x10
[Tue Sep 26 14:00:14 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Tue Sep 26 14:00:14 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=3026485,uid=1000
[Tue Sep 26 14:00:14 2023] Out of memory: Killed process 3026485 (postgres) total-vm:21820372kB, anon-rss:10458096kB, file-rss:812kB, shmem-rss:4570316kB, UID:1000 pgtables:33532kB oom_score_adj:0
[Tue Sep 26 14:06:17 2023] gdbus invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Tue Sep 26 14:06:17 2023]  oom_kill_process.cold+0xb/0x10
[Tue Sep 26 14:06:17 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Tue Sep 26 14:06:17 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=fwupd-refresh.service,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=3027540,uid=1000
[Tue Sep 26 14:06:17 2023] Out of memory: Killed process 3027540 (postgres) total-vm:24653208kB, anon-rss:14573012kB, file-rss:1440kB, shmem-rss:516660kB, UID:1000 pgtables:31116kB oom_score_adj:0
[Tue Sep 26 14:13:54 2023] postgres invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Tue Sep 26 14:13:54 2023]  oom_kill_process.cold+0xb/0x10
[Tue Sep 26 14:13:54 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Tue Sep 26 14:13:54 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=3028609,uid=1000
[Tue Sep 26 14:13:54 2023] Out of memory: Killed process 3028609 (postgres) total-vm:21351752kB, anon-rss:9560036kB, file-rss:936kB, shmem-rss:5482876kB, UID:1000 pgtables:35008kB oom_score_adj:0
[Tue Sep 26 16:48:55 2023] bash invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Tue Sep 26 16:48:55 2023]  oom_kill_process.cold+0xb/0x10
[Tue Sep 26 16:48:55 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Tue Sep 26 16:48:55 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=3035495,uid=1000
[Tue Sep 26 16:48:55 2023] Out of memory: Killed process 3035495 (postgres) total-vm:21856552kB, anon-rss:10700616kB, file-rss:660kB, shmem-rss:4462576kB, UID:1000 pgtables:33592kB oom_score_adj:0
[Tue Sep 26 16:54:35 2023] postgres invoked oom-killer: gfp_mask=0x1100cca(GFP_HIGHUSER_MOVABLE), order=0, oom_score_adj=0
[Tue Sep 26 16:54:35 2023]  oom_kill_process.cold+0xb/0x10
[Tue Sep 26 16:54:35 2023] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Tue Sep 26 16:54:35 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=user.slice,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/session-2.scope,task=postgres,pid=3035770,uid=1000
[Tue Sep 26 16:54:35 2023] Out of memory: Killed process 3035770 (postgres) total-vm:21849260kB, anon-rss:10405316kB, file-rss:808kB, shmem-rss:4589612kB, UID:1000 pgtables:33572kB oom_score_adj:0
```

**Comment by mdcallag** [comment]
Example error messages in logfile when the OOM killer arrives

```
... Failed process was running: SELECT SUM(k) FROM sbtest1 WHERE id BETWEEN $1 AND $2
... Failed process was running:  SELECT k FROM sbtest1 WHERE k BETWEEN $1 AND $2
... Failed process was running: SELECT c FROM sbtest1 WHERE id BETWEEN $1 AND $2
```

The failed query for round 4 was a point query:
```
... LOG:  server process (PID 3036874) was terminated by signal 9: Killed
... DETAIL:  Failed process was running: SELECT c FROM sbtest1 WHERE id=11067451
```

**Comment by mdcallag** [comment]
From valgrind --tool=massif I see this in a setup similar to Round 4 above. So I assume the OOM is from memory allocated in one of the 3 places listed here.

```
99.90% (4,363,880,140B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->99.87% (4,362,532,864B) 0x6FCACB: AllocSetContextCreateInternal (aset.c:438)
| ->33.29% (1,454,014,464B) 0x26A2EA: printtup_startup (printtup.c:128)
| | ->33.29% (1,454,014,464B) 0x3FE855: standard_ExecutorRun (execMain.c:354)
| |   ->33.29% (1,454,014,464B) 0x765F114: orioledb_ExecutorRun_hook (ddl.c:1336)
| |     ->33.29% (1,454,014,464B) 0x5A230D: PortalRunSelect (pquery.c:924)
| |       ->33.29% (1,454,014,464B) 0x5A37F7: PortalRun (pquery.c:768)
| |         ->33.29% (1,454,014,464B) 0x59FB2B: exec_simple_query (postgres.c:1274)
| |           ->33.29% (1,454,014,464B) 0x5A1D05: PostgresMain (postgres.c:4637)
| |             ->33.29% (1,454,014,464B) 0x518931: BackendRun (postmaster.c:4478)
| |               ->33.29% (1,454,014,464B) 0x518931: BackendStartup (postmaster.c:4206)
| |                 ->33.29% (1,454,014,464B) 0x518931: ServerLoop (postmaster.c:1786)
| |                   ->33.29% (1,454,014,464B) 0x519903: PostmasterMain (postmaster.c:1470)
| |                     ->33.29% (1,454,014,464B) 0x25A585: main (main.c:198)
| |
| ->33.29% (1,454,014,464B) 0x40AFE1: CreateExprContextInternal.constprop.0 (execUtils.c:261)
| | ->33.29% (1,454,014,464B) 0x40B5E9: CreateExprContext (execUtils.c:311)
| |   ->33.29% (1,454,014,464B) 0x40B5E9: ExecAssignExprContext (execUtils.c:490)
| |     ->33.29% (1,454,014,464B) 0x41732A: ExecInitCustomScan (nodeCustom.c:56)
| |       ->33.29% (1,454,014,464B) 0x405535: ExecInitNode (execProcnode.c:290)
| |         ->33.29% (1,454,014,464B) 0x3FF172: InitPlan (execMain.c:974)
| |           ->33.29% (1,454,014,464B) 0x3FF172: standard_ExecutorStart (execMain.c:266)
| |             ->33.29% (1,454,014,464B) 0x5A2FD5: PortalStart (pquery.c:517)
| |               ->33.29% (1,454,014,464B) 0x59FAAA: exec_simple_query (postgres.c:1235)
| |                 ->33.29% (1,454,014,464B) 0x5A1D05: PostgresMain (postgres.c:4637)
| |                   ->33.29% (1,454,014,464B) 0x518931: BackendRun (postmaster.c:4478)
| |                     ->33.29% (1,454,014,464B) 0x518931: BackendStartup (postmaster.c:4206)
| |                       ->33.29% (1,454,014,464B) 0x518931: ServerLoop (postmaster.c:1786)
| |                         ->33.29% (1,454,014,464B) 0x519903: PostmasterMain (postmaster.c:1470)
| |                           ->33.29% (1,454,014,464B) 0x25A585: main (main.c:198)
| |
| ->33.29% (1,454,014,464B) 0x769743A: o_begin_custom_scan (scan.c:600)
| | ->33.29% (1,454,014,464B) 0x4173B2: ExecInitCustomScan (nodeCustom.c:111)
| |   ->33.29% (1,454,014,464B) 0x405535: ExecInitNode (execProcnode.c:290)
| |     ->33.29% (1,454,014,464B) 0x3FF172: InitPlan (execMain.c:974)
| |       ->33.29% (1,454,014,464B) 0x3FF172: standard_ExecutorStart (execMain.c:266)
| |         ->33.29% (1,454,014,464B) 0x5A2FD5: PortalStart (pquery.c:517)
| |           ->33.29% (1,454,014,464B) 0x59FAAA: exec_simple_query (postgres.c:1235)
| |             ->33.29% (1,454,014,464B) 0x5A1D05: PostgresMain (postgres.c:4637)
| |               ->33.29% (1,454,014,464B) 0x518931: BackendRun (postmaster.c:4478)
| |                 ->33.29% (1,454,014,464B) 0x518931: BackendStartup (postmaster.c:4206)
| |                   ->33.29% (1,454,014,464B) 0x518931: ServerLoop (postmaster.c:1786)
| |                     ->33.29% (1,454,014,464B) 0x519903: PostmasterMain (postmaster.c:1470)
| |                       ->33.29% (1,454,014,464B) 0x25A585: main (main.c:198)
```

**Comment by mdcallag** [comment]
For the Round 4 query that gets OOM I ran it for less time and got valgrind --leak-check=full output

There were 3 call stacks with large amounts for still reachable. And these might match the 3 call stacks above from massic.

```
==00:00:04:00.150 3045703== 202,596,352 bytes in 24,731 blocks are still reachable in loss record 137 of 139
==00:00:04:00.150 3045703==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==00:00:04:00.150 3045703==    by 0x6FCACB: AllocSetContextCreateInternal (aset.c:438)
==00:00:04:00.150 3045703==    by 0x40AFE1: CreateExprContextInternal.constprop.0 (execUtils.c:261)
==00:00:04:00.150 3045703==    by 0x40B5E9: CreateExprContext (execUtils.c:311)
==00:00:04:00.150 3045703==    by 0x40B5E9: ExecAssignExprContext (execUtils.c:490)
==00:00:04:00.150 3045703==    by 0x41732A: ExecInitCustomScan (nodeCustom.c:56)
==00:00:04:00.150 3045703==    by 0x405535: ExecInitNode (execProcnode.c:290)
==00:00:04:00.150 3045703==    by 0x3FF172: InitPlan (execMain.c:974)
==00:00:04:00.150 3045703==    by 0x3FF172: standard_ExecutorStart (execMain.c:266)
==00:00:04:00.150 3045703==    by 0x5A2FD5: PortalStart (pquery.c:517)
==00:00:04:00.150 3045703==    by 0x59FAAA: exec_simple_query (postgres.c:1235)
==00:00:04:00.150 3045703==    by 0x5A1D05: PostgresMain (postgres.c:4637)
==00:00:04:00.150 3045703==    by 0x518931: BackendRun (postmaster.c:4478)
==00:00:04:00.150 3045703==    by 0x518931: BackendStartup (postmaster.c:4206)
==00:00:04:00.150 3045703==    by 0x518931: ServerLoop (postmaster.c:1786)
==00:00:04:00.150 3045703==    by 0x519903: PostmasterMain (postmaster.c:1470)
```

```
==00:00:04:00.150 3045703== 202,596,352 bytes in 24,731 blocks are still reachable in loss record 138 of 139
==00:00:04:00.150 3045703==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==00:00:04:00.150 3045703==    by 0x6FCACB: AllocSetContextCreateInternal (aset.c:438)
==00:00:04:00.150 3045703==    by 0x76A143A: o_begin_custom_scan (scan.c:600)
==00:00:04:00.150 3045703==    by 0x4173B2: ExecInitCustomScan (nodeCustom.c:111)
==00:00:04:00.150 3045703==    by 0x405535: ExecInitNode (execProcnode.c:290)
==00:00:04:00.150 3045703==    by 0x3FF172: InitPlan (execMain.c:974)
==00:00:04:00.150 3045703==    by 0x3FF172: standard_ExecutorStart (execMain.c:266)
==00:00:04:00.150 3045703==    by 0x5A2FD5: PortalStart (pquery.c:517)
==00:00:04:00.150 3045703==    by 0x59FAAA: exec_simple_query (postgres.c:1235)
==00:00:04:00.150 3045703==    by 0x5A1D05: PostgresMain (postgres.c:4637)
==00:00:04:00.150 3045703==    by 0x518931: BackendRun (postmaster.c:4478)
==00:00:04:00.150 3045703==    by 0x518931: BackendStartup (postmaster.c:4206)
==00:00:04:00.150 3045703==    by 0x518931: ServerLoop (postmaster.c:1786)
==00:00:04:00.150 3045703==    by 0x519903: PostmasterMain (postmaster.c:1470)
==00:00:04:00.150 3045703==    by 0x25A585: main (main.c:198)
```

```
==00:00:04:00.150 3045703== 202,596,352 bytes in 24,731 blocks are still reachable in loss record 139 of 139
==00:00:04:00.150 3045703==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==00:00:04:00.150 3045703==    by 0x6FCACB: AllocSetContextCreateInternal (aset.c:438)
==00:00:04:00.150 3045703==    by 0x26A2EA: printtup_startup (printtup.c:128)
==00:00:04:00.150 3045703==    by 0x3FE855: standard_ExecutorRun (execMain.c:354)
==00:00:04:00.150 3045703==    by 0x7669114: orioledb_ExecutorRun_hook (ddl.c:1336)
==00:00:04:00.150 3045703==    by 0x5A230D: PortalRunSelect (pquery.c:924)
==00:00:04:00.150 3045703==    by 0x5A37F7: PortalRun (pquery.c:768)
==00:00:04:00.150 3045703==    by 0x59FB2B: exec_simple_query (postgres.c:1274)
==00:00:04:00.150 3045703==    by 0x5A1D05: PostgresMain (postgres.c:4637)
==00:00:04:00.150 3045703==    by 0x518931: BackendRun (postmaster.c:4478)
==00:00:04:00.150 3045703==    by 0x518931: BackendStartup (postmaster.c:4206)
==00:00:04:00.150 3045703==    by 0x518931: ServerLoop (postmaster.c:1786)
==00:00:04:00.150 3045703==    by 0x519903: PostmasterMain (postmaster.c:1470)
==00:00:04:00.150 3045703==    by 0x25A585: main (main.c:198)
```

**Comment by mdcallag** [comment]
Confirmed I can reproduce this. It requires:
* range queries, such as SELECT SUM(k) FROM sbtest1 WHERE id BETWEEN $1 AND $2
* vacuum and/or checkpoint

When the range queries run before vacuum+checkpoint there is no OOM.
When they run after there is OOM.

From your github page there is this but I don't know if that means it should not be used or it doesn't have to be used. Do I need to do vacuum? Do I need to do checkpoint? Should I just run analyze to collect stats? How do stats get updated over time, because vacuum can help with that with upstream Postgres.
```
Dedicated VACUUMing of tables is not needed as well, removing a significant and common cause of system performance deterioration and database outages.
```

A reproduction script
```
echo Load at $( date )
/home/mdcallag/sysb/bin/sysbench --db-driver=pgsql --range-size=100 --table-size=20000000 --tables=1 --events=0 --time=600 /home/mdcallag/sysb/share/sysbench/oltp_point_select.lua prepare --pgsql-user=root --pgsql-password=pw --pgsql-host=127.0.0.1 --pgsql-db=ib --create-table-options="using orioledb" >& o.sb.ld

sleep 10

# This does not get OOM
while :; do date; ps aux | grep postgres | grep -v python | grep -v psql | grep -v grep; sleep 1; done >& o.ps.r1 &
pspid=$!
echo Range queries, round 1 at $( date )
/home/mdcallag/sysb/bin/sysbench --db-driver=pgsql --range-size=10 --table-size=20000000 --tables=1 --threads=1 --events=0 --warmup-time=5 --time=600 /home/mdcallag/sysb/share/sysbench/oltp_read_only.lua run --pgsql-user=root --pgsql-password=pw --pgsql-host=127.0.0.1 --pgsql-db=ib --rand-type=uniform --create-table-options="using orioledb" >& o.sb.r1
kill $pspid

echo Vacuum at $( date )
/home/mdcallag/d/pg16oriole_o2_nofp_patches16_18/bin/psql -Uroot -h127.0.0.1 ib -c "vacuum (analyze, verbose) sbtest1"
echo Checkpoint at $( date )
/home/mdcallag/d/pg16oriole_o2_nofp_patches16_18/bin/psql -Uroot -h127.0.0.1 ib -c "checkpoint"
echo Sleep for 180s at $( date )
sleep 180

# This gets OOM
while :; do date; ps aux | grep postgres | grep -v python | grep -v psql | grep -v grep; sleep 1; done >& o.ps.r2 &
pspid=$!
echo Range queries, round 2 at $( date )
/home/mdcallag/sysb/bin/sysbench --db-driver=pgsql --range-size=10 --table-size=20000000 --tables=1 --threads=1 --events=0 --warmup-time=5 --time=600 /home/mdcallag/sysb/share/sysbench/oltp_read_only.lua run --pgsql-user=root --pgsql-password=pw --pgsql-host=127.0.0.1 --pgsql-db=ib --rand-type=uniform --create-table-options="using orioledb" >& o.sb.r2

kill $pspid

```

**Comment by mdcallag** [comment]
I was too optimistic about the OOM not happening when vacuum is not done. It still happens. I am running more tests to see if it can be avoided.

**Comment by mdcallag** [comment]
Repeated the test where the parts that do queries run for more time (3600 seconds) and I collect output from "ps" at one second intervals. And with that there is a memory leak during range queries run both before and after the analyze, checkpoint, vacuum steps. 

For the numbers below the values of (VSZ, RSS) during point queries is about (9G, 4.5G). The OOM mention below occurs because (VSZ, RSS) during range queries reaches (21.1G, 14.2G).

One example, from the range queries run before analyze, checkpoint and vacuum, the values of VSZ and RSS from "ps" as GB from the first 20 seconds of the test
```
VSZ. RSS
8.8     0.8
8.8     1.5
8.9     2.1
9.0     2.6
9.1     3.1
9.1     3.4
9.2     3.8
9.3     4.0
9.4     4.3
9.5     4.5
9.5     4.7
9.6     4.9
9.7     5.0
9.8     5.2
9.9     5.3
9.9     5.4
10.0    5.6
10.1    5.7
10.2    5.8
10.2    5.9
```

And then from the last 20 seconds just prior to the OOM killer arriving
```
VSZ. RSS
21.0    14.2
21.0    14.2
21.0    14.2
21.0    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
21.1    14.2
```

**Comment by pashkinelfe** [comment]
https://github.com/orioledb/orioledb/commit/993070d0cbcc321321f6dcee505c8788a231a0af fixes memory leak in #282. @mdcallag could you recheck your scenario with this fix?

**Comment by mdcallag** [comment]
Recompiled with the patch, restarted tests and it appears to have fixed the leak. Will know for sure in one day, but for now I think it is OK to close this as fixed.
